src/crowdin_client.cpp

```diff
@@ -50,64 +57,71 @@
+#define OAUTH_CLIENT_ID     "k0uFz5HYQh0VzWgZmOpA"
+//      any arbitrary unique unguessable string (e.g. UUID in hex)
+#define OAUTH_STATE         "948cf13ffffb47119d6cfa2b68898f67"
+//      The value of below macro should be set exactly as is (without quotes)
+//      to "Authorization Callback URL" of Crowdin application
+//      https://support.crowdin.com/enterprise/creating-oauth-app/
 #define OAUTH_URI_PREFIX    "poedit://auth/crowdin/"
-
-// Recursive extract files from /api/project/*/info response
-void ExtractFilesFromInfo(std::vector<std::wstring>& out, const json& r, const std::wstring& prefix)
-{
-    for (auto i : r["files"])
-    {
-        std::wstring name = prefix + str::to_wstring(i["name"]);
-        std::string node_type = i["node_type"];
-        if (node_type == "file")
-        {
-            out.push_back(name);
-        }
-        else if (node_type == "directory" || node_type == "branch")
-        {
-            ExtractFilesFromInfo(out, i, name + L"/");
-        }
-    }
-}
+#define OAUTH_AUTHORIZE_URL "/oauth/authorize" \
+    "?" "response_type" "=" "token" \
+    "&" "scope"         "=" OAUTH_SCOPE \
+    "&" "client_id"     "=" OAUTH_CLIENT_ID \
+    "&" "state"         "=" OAUTH_STATE \
+    "&" "redirect_uri"  "=" OAUTH_URI_PREFIX \
+    "&" "utm_source=poedit.net&utm_medium=referral&utm_campaign=poedit" // source tracking
 
 } // anonymous namespace
 
-
 std::string CrowdinClient::WrapLink(const std::string& page)
 {
-    std::string url("https://poedit.net/crowdin");
+    std::string url("https://accounts.crowdin.com");
     if (!page.empty() && page != "/")
-        url += "?u=" + http_client::url_encode(page);
+        url += page;
     return url;
 }
 
 
 class CrowdinClient::crowdin_http_client : public http_client
 {
 public:
-    crowdin_http_client(CrowdinClient& owner)
-        : http_client("https://api.crowdin.com"), m_owner(owner)
+    crowdin_http_client(CrowdinClient& owner, const std::string& url_prefix)
+        : http_client(url_prefix), m_owner(owner)
     {}
 
 protected:
     std::string parse_json_error(const json& response) const override
     {
-        std::string msg = response["error"]["message"];
-
-        // Translate commonly encountered messages:
-        if (msg == "Translations download is forbidden by project owner")
-            msg = _("Downloading translations is disabled in this project.").utf8_str();
-
-        return msg;
+        wxLogTrace("poedit.crowdin", "JSON error: %s", response.dump().c_str());
+        
+        try
+        {
+            return response.at("errors").at(0).at("error").at("errors").at(0).at("message");
+        }
+        catch(...)
+        {
+            try
+            {
+                return response.at("error").at("message");
+            }
+            catch (...)
+            {
+                std::string msg;
+                msg = _("JSON request error").utf8_str();
+                return msg;
+            }
+        }
```

Comment:
Can you please include a comment either showing expected JSON structure of the reply, or linking to API docs showing it? I still find this throw-and-catch cascade opaque and hard to understand, a comment showing the data it works with would help.

In particular I am confused by there being more `errors` nested inside a singular `error` which is itself contained in a possible list of errors… Does the Crowdin API really return *list of errors that themselves contain multitude of errors*? 😳

And if multiple errors can be returned, shouldn't they be shown to the user? 

Or use the [contains()](https://github.com/nlohmann/json/pull/1474) function without throwing, but I can see how the cascade is actually less complicated to write...

Reply:
Commented 278a08250a71def13b71847c9e962309d829e90e

> In particular I am confused by there being more errors nested inside a singular error which is itself contained in a possible list of errors… Does the Crowdin API really return list of errors that themselves contain multitude of errors?

I don't know what how to explain/formalize this and what the purpose purpose is it's done in such way. But just working with handling the fact :). 

> And if multiple errors can be returned, shouldn't they be shown to the user?

I never encountered it, at least while looked into responses with such structure returned by all API requests used in Poedit. But you think it's so important to implement displaying of "all errors" just for that rare potential or may be even pure hypothetical case they will be there - please unresolve this comment so I will implement or at least check with Crowdin API team whether it's really the case worth implementing

Reply:
> But you think it's so important to implement displaying of "all errors"

Nah, I was just trying to understand why what seems like convoluted extraction of errors takes place… Let's shrug it off as c'est la vie...